### DISCUSSION TITLE
[@types/react] ComponentProps generic does not fallback to `{}` when inferring props from a function with no arguments

### DISCUSSION THREAD

**Discussion by Yelmuf** [DefinitelyTyped/DefinitelyTyped#63503] [Issues with a @types package]
Hi! I have a problem with a fallback when inferring types with `React.ComponentProps` from a function with no arguments.

## Problem

Consider a component that is a function without props:
```tsx
import React from 'react'

function Component() {
    return <></>
}

type Props = React.ComponentProps<typeof Component>
//   ^? type Props = unknown
```

**Question:** shouldn't it be `type Props = {}`?

This behavior can be overcome by explicit typing of a component:

```tsx
import React from 'react'

const ExplicitComponent: React.FC = () => {
    return <></>
}

type Props = React.ComponentProps<typeof ExplicitComponent>
//   ^? type Props = {}
```

Since it's [discouraged](https://github.com/facebook/create-react-app/pull/8177) to actually use `React.FC` type in codebase, I am confused with the abovementioned behavior of `React.ComponentProps`.

Versions in `package.json`:
```json
 "@types/react": "18.0.25",
 "typescript": "4.7.4",
```

## Possible solution

`ComponentProps` has a "final" fallback `{}`, but in case of `T = () => JSX.Element` `P` will be inferred as `unknown`. Suggested change would result in `{}` in case of absence of props.
```diff
type ComponentProps<T extends keyof JSX.IntrinsicElements | JSXElementConstructor<any>> =
    T extends JSXElementConstructor<infer P>
-       ? P
+       ? P & {}
        : T extends keyof JSX.IntrinsicElements
            ? JSX.IntrinsicElements[T]
            : {};
```

What do you think?

[a comment by an automated account is not shown]

**Comment by eps1lon** [comment]
`{}` would be problematic since that type means "any non-nullish type" i.e. `number`, `string` etc would be fine to pass as props. An empty objecty would be fine I imagine (not `object` since that would allow any props).

Would appreciate if you can write some tests/implementation in PR form to check how compatible this is with existing types.
